### ISSUE TITLE
Empty response body causes error or incorrect content-type inference/validation

### ISSUE THREAD

**Issue by rparini** [spec-first/connexion#1850]
### Description

If an endpoint returns a response with no body without specifying the content-type and the API spec for the endpoint has other, non-empty, responses as options then Connexion errors if those responses contain multiple other content types. If the non-empty responses only have one content-type then Connexion will not error but will return a response using that one content-type as the `content-type` header for the empty response, which may not be appropriate.

If instead the application does explicitly specify a suitable content type or relies on the default content-type (for the `FlaskApp` that's `text/html`) and `validate_responses=True` then Connexion's response validation throws an error if that content-type does not appear in one of the non-empty responses in the schema.

Looks related to https://github.com/spec-first/connexion/issues/1847 although that ticket focuses on the warnings raised while my issue is Connexion's content-type inference and response validation behaviour. But can close this and post a comment there instead if it's felt that's more appropriate.

### Expected behaviour

With the openapi schema
```yaml
# api.yaml
openapi: 3.0.0
info:
  title: Sample API
  version: 0.1.0
paths:
  /empty:
    get:
      operationId: app.empty_get
      security: []
      description: Empty
      responses:
        '204':
          description: No content
        '200':
          description: Image
          content:
            image/png:
             schema: 
               type: string
               format: binary
        '400':
          description: Bad request
          content:
            application/json:
              schema:
                type: object
                properties:
                  description:
                    type: string

```
and the application
```python
# app.py
from pathlib import Path
import connexion

app = connexion.FlaskApp(__name__, validate_responses=True)
app.add_api("api.yaml")

def empty_get():
    return None

if __name__ == "__main__":
    app.run(f"{Path(__file__).stem}:app")
```

I would expect calling the `/empty` endpoint to return a response with a 204 status code and no `content-type` header (or, debatably, an appropriate `content-type` that makes sense for an empty body like `content-type: text/plain`, not sure what the most "correct" behaviour is)

### Actual behaviour
calling the /empty endpoint errors with:
```bash
>>> curl http://127.0.0.1:8000/empty
{
  "type": "about:blank", 
  "title": "Internal Server Error", 
  "detail": "Multiple response content types are defined in the operation spec, but the handler response did not specify which one to return.", 
  "status": 500
}
```
even though the application's controller returns `None` as the [documentation suggests](https://connexion.readthedocs.io/en/latest/response.html#status-code) for empty responses. Same behaviour if instead `connexion.datastructures.NoContent` is returned by the controller.

If the `400` response is removed from the schema then the endpoint does not error but it returns a response with the `content-type: image/png` since that's the only non-empty content-type left in the spec.
```
>>> curl http://127.0.0.1:8000/empty -v
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000
> GET /empty HTTP/1.1
> Host: 127.0.0.1:8000
> User-Agent: curl/8.4.0
> Accept: */*
> 
< HTTP/1.1 204 No Content
< date: Sat, 13 Jan 2024 22:00:03 GMT
< server: uvicorn
< content-type: image/png
< 
* Connection #0 to host 127.0.0.1 left intact
```

If I instead change the application to explicitly return a suitable content-type like `"text/plain"` but with `validate_responses=True`:
```python
def empty_get():
    return flask.Response(status=204, content_type="text/plain")
```
then the application will error if `"text/plain"` is not one of the non-empty content-types specified in the schema:
```json
{
  "type": "about:blank", 
  "title": "Internal Server Error", 
  "detail": "Invalid Response Content-type (text/plain), expected ['image/png', 'application/json']", 
  "status": 500
}
```
I would have to set the content-type to `image/png` or `application/json` to pass the validation but these are IMO not suitable content-types for an empty response.

I also tried to get the `content-type` header to not appear in the response at all but this does not seem to be possible. Setting `flask.Response(status=204, content_type=None)` or `connexion.lifecycle.ConnexionResponse(status_code=204, content_type=None)` just means the content-type gets inferred to `text/html` which then errors at response validation because `text/html` is not one of the non-empty content-types in the schema.

### Additional info:

Output of the commands:

- `python --version`: Python 3.9.13
- `pip show connexion | grep "^Version\:"`: Version: 3.0.5

**Comment by Gidgidonihah** [comment]
I came to create an issue, but this is *very* related to what I was going to post.

**Shouldn't content-type be delineated by status code?**

Given the example here, if you're returning `None`, or `connexion.datastructures.NoContent` connexion will convert that to a 204. There aren't multiple content types defined for that status code.

Likewise, if you return a 200 status code, the only valid response code is `image/png`. Shouldn't connexion infer `image/png`?
Why should it matter that you defined `application/json` for status code 400?

**Comment by RobbeSneyders** [comment]
Thanks for the detailed issue @Gidgidonihah 

Inferring the content type correctly in all cases is not straightforward, especially due to the difference in response definition between the Swagger 2 and OpenAPI 3 specs. We implemented a simple heuristic which can be overridden by explicitly defining the content type both in the response and the specification. I agree that this is not ideal for a response without content and that the heuristic in general can be improved though.

Any improvements to this part of the code base are welcome.

**Comment by Gidgidonihah** [comment]
Indeed, this makes sense given that v2 doesn't define the content type by status code, but rather at the response level.

We would then obviously have to differ the inference based upon the schema version. 

For v2, there's not much that could be done, but if using a v3 schema, it seems like it would be fairly straightforward to modify [`_infer_content_type`](https://github.com/spec-first/connexion/blob/main/connexion/decorators/response.py#L40) to take an optional status code and, when present, extract the possible content types from `operation.responses` using the `status_code`, rather than from `operation.produces`. Then the question becomes how do you branch the logic? IMO, I'd modify `AbstractOperation` to take an optional `status_code` on the `produces` method. `Swagger2Operation` could simply ignore it, while `OpenAPIOperation` would return only the content types from self._responses when status_code is present. 

If that seems like a reasonable approach, I'd be happy to submit a PR for it.

**Comment by FlorianIkarus** [comment]
I just stumbled across this issue, which very much summarizes my problem after upgrading to Connexion 3.

For one endpoint in my OpenAPI v3 specification, I have defined `text/html` for a successful response and `application/json` for any other (error) response. I wondered why the content type was not inferred correctly from the specification. So I started digging around and found `_infer_content_type`, which made it clear for me that the status code is not used at all.

This means that I have to set the content type excplicitly everywhere, because I cannot guarantee that a response returned by a library (e.g. default 500 handler) is not used in a "mixed" context.

I wasn't aware of the v2 / v3 inference difficulties, so I understand the simple heuristic approach. However, I really like @Gidgidonihah's suggestion to handle inference differently based on v2 / v3.
